gusto/equations.py

```diff
@@ -0,0 +1,388 @@
+        trials = TrialFunctions(W)
+        X = Function(W)
+        u, D = X.split()
+
+        # define velocity advection term
+        if self.u_advection_option == "vector_invariant_form":
+            u_adv = vector_invariant_form(state, W, 0)
+        elif self.u_advection_option == "vector_advection":
+            u_adv = advection_vector_manifold_form(state, W, 0)
+        elif self.u_advection_option == "circulation_form":
+            ke_form = kinetic_energy_form(state, W, 0)
+            ke_form = advection.remove(ke_form)
+            ke_form = ke_form.label_map(all_terms,
+                                        replace_labelled(subject,
+                                                         advecting_velocity))
+            u_adv = advection_equation_circulation_form(state, W, 0) + ke_form
+        else:
+            raise ValueError("Invalid u_advection_option: %s" % self.u_advection_option)
+
+        # define pressure gradient term and its linearisation
+        pressure_gradient_term = subject(-g*div(w)*D*dx, X)
+        linear_pg_term = pressure_gradient_term.label_map(
+            all_terms, replace_labelled(trials, subject))
+
+        # the base form for u contains the velocity advection term and
+        # the pressure gradient term
+        u_form = u_adv + linearisation(pressure_gradient_term, linear_pg_term)
+
+        # add any additional terms
+        for field_name in ["coriolis", "topography"]:
+            try:
+                field = state.fields(field_name)
+                add_term = True
+            except AttributeError:
+                logger.info("field %s not present" % field_name)
+                add_term = False
+
+            if add_term:
+                if field_name == "coriolis":
+                    # define the coriolis term and its linearisation
+                    coriolis_term = subject(field*inner(w, state.perp(u))*dx, X)
+                    linear_coriolis_term = coriolis_term.label_map(
+                        all_terms, replace_labelled(trials, subject))
+                    # add on the coriolis term
+                    u_form += linearisation(coriolis_term, linear_coriolis_term)
+
+                elif field_name == "topography":
+                    # add on the topography term - the linearisation
+                    # is not defined as we don't usually make it part
+                    # of the linear solver, However, this will have to
+                    # be defined when we start using exponential
+                    # integrators.
+                    u_form += -g*div(w)*field*dx
+
+        # define the depth continuity term and its linearisation
+        Dadv = continuity_form(state, W, 1)
+        Dadv_linear = linear_continuity_form(state, W, 1, qbar=H).label_map(
+            all_terms, replace_labelled(trials, subject, advecting_velocity))
+        D_form = linearisation(Dadv, Dadv_linear)
+
+        return index(u_form, 0) + index(D_form, 1)
```

Comment:
This `index` label is possibly not necessary but maybe I don't know enough ufl. I want to be able to extract the part of the form that corresponds to the `u` equation or the `D` equation. I thought I would be able to do this by examining the function space of the test function but I couldn't figure out a way - any thoughts?